packages/components/src/unit-control/unit-select-control.tsx

```diff
@@ -56,7 +59,7 @@ export default function UnitSelectControl( {
 			className={ classes }
 			onChange={ handleOnChange }
 			size={ size }
```

Comment:
@ciampo I've made a start on refactoring the rest of the UnitControl to TypeScript, and am working through looking at a couple of different examples we've already got of TS usage in the SelectControl and InputControl components.

I noticed that this line complains about `size` being unassignable to type `never`. I noticed that the SelectControl uses a `selectSize` type instead. I can easily rename it, but I was wondering if you know why this might be complaining?

![image](https://user-images.githubusercontent.com/14988353/135578463-442f9c88-3d68-4a98-bd5d-437c2bf94a8c.png)

I also started importing / inheriting from some of the types in InputControl, but then wondered if it's actually better to create new types in this component to keep them isolated? Curious to hear your opinion on how DRY we need to make component types.

No rush at all on feedback here, though — I'm wrapping up for the week, but will chip away at this next week, too. Thanks!

Reply:
This took a while to debug.

I started by being even more specific about the props that `UnitSelect` supports:

```ts
// in styles.ts
export const UnitSelect = styled.select< UnitControlLabelProps >`
// styles ...
`;
```

as suggested in the [Emotion docs](https://emotion.sh/docs/typescript#passing-props). But it  didn't fix the issue.

And so I started looking deeper into it. The `never` type represents something that can never occur — so why is that?

Turns out, `size` is being declared twice:

1. since `UnitSelect` is a `styled.select`, it inherits all of the props of the `select` HTML element (`React.SelectHTMLAttributes<HTMLSelectElement>`). One of these happens to be [the `size` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#attr-size). This attribute is typed as a `number`
2. At the same time, we're also telling TypeScript that this component accepts props of type `UnitControlLabelProps`. In those props, there's a prop named `size` which can be `'default' | 'small'`

Since TypeScript can't find a valid intersection of the 2 types (`number` and `'default' | 'small'`), the result is that the `size` prop for `UnitSelect` gets a type of `never`.

> I noticed that the SelectControl uses a `selectSize` type instead

Now we know why! I went ahead and made similar changes, and this seems to make TypeScript happy:

<details><summary>click to expand</summary>

```diff
diff --git a/packages/components/src/unit-control/styles/unit-control-styles.ts b/packages/components/src/unit-control/styles/unit-control-styles.ts
index a85288dfc5..05b60415d6 100644
--- a/packages/components/src/unit-control/styles/unit-control-styles.ts
+++ b/packages/components/src/unit-control/styles/unit-control-styles.ts
@@ -8,7 +8,11 @@ import styled from '@emotion/styled';
  */
 import { COLORS, rtl } from '../../utils';
 import NumberControl from '../../number-control';
-import type { UnitControlLabelProps } from '../types';
+import type { SelectSize } from '../types';
+
+// Using `selectSize` instead of `size` to avoid a type conflict with the
+// `size` HTML attribute of the `select` element.
+type SelectProps = {
+	selectSize: SelectSize;
+};
 
 export const Root = styled.div`
 	box-sizing: border-box;
@@ -52,7 +56,7 @@ export const ValueInput = styled( NumberControl )`
 	}
 `;
 
-const unitSizeStyles = ( { size }: UnitControlLabelProps ) => {
+const unitSizeStyles = ( { selectSize }: SelectProps ) => {
 	const sizes = {
 		default: {
 			height: 28,
@@ -68,10 +72,10 @@ const unitSizeStyles = ( { size }: UnitControlLabelProps ) => {
 		},
 	};
 
-	return css( sizes[ size ] );
+	return css( sizes[ selectSize ] );
 };
 
-const baseUnitLabelStyles = ( props: UnitControlLabelProps ) => {
+const baseUnitLabelStyles = ( props: SelectProps ) => {
 	return css`
 		appearance: none;
 		background: transparent;
@@ -96,7 +100,7 @@ const baseUnitLabelStyles = ( props: UnitControlLabelProps ) => {
 	`;
 };
 
-export const UnitLabel = styled.div`
+export const UnitLabel = styled.div< SelectProps >`
 	&&& {
 		pointer-events: none;
 
@@ -104,7 +108,7 @@ export const UnitLabel = styled.div`
 	}
 `;
 
-export const UnitSelect = styled.select`
+export const UnitSelect = styled.select< SelectProps >`
 	&&& {
 		${ baseUnitLabelStyles };
 		cursor: pointer;
diff --git a/packages/components/src/unit-control/types.ts b/packages/components/src/unit-control/types.ts
index 8acc7d5426..21aae0fa86 100644
--- a/packages/components/src/unit-control/types.ts
+++ b/packages/components/src/unit-control/types.ts
@@ -11,7 +11,7 @@ import type { InputControlLabelProps } from '../input-control/types';
 
 export type Value = number | string;
 
-export type Size = 'default' | 'small';
+export type SelectSize = 'default' | 'small';
 
 export type WPUnitControlUnit = {
 	/**
@@ -51,6 +51,6 @@ export type UnitSelectControlProps = {
 			data: WPUnitControlUnit;
 		}
 	) => void;
-	size?: Size;
+	size?: SelectSize;
 	value: string;
 };
diff --git a/packages/components/src/unit-control/unit-select-control.tsx b/packages/components/src/unit-control/unit-select-control.tsx
index d1f38ecc8b..6701f54fe3 100644
--- a/packages/components/src/unit-control/unit-select-control.tsx
+++ b/packages/components/src/unit-control/unit-select-control.tsx
@@ -38,7 +38,7 @@ export default function UnitSelectControl( {
 		return (
 			<UnitLabel
 				className="components-unit-control__unit-label"
-				size={ size }
+				selectSize={ size }
 			>
 				{ value }
 			</UnitLabel>
@@ -58,7 +58,7 @@ export default function UnitSelectControl( {
 		<UnitSelect
 			className={ classes }
 			onChange={ handleOnChange }
-			size={ size }
+			selectSize={ size }
 			tabIndex={ isTabbable ? null : -1 }
 			value={ value }
 			{ ...props }
```
</details>

Reply:
Thanks so much for investigating, @ciampo and for your detailed notes! That makes perfect sense now, and it's a great reminder to check the HTML spec in future when I encounter conflicts with HTML elements 😃

I've tested and committed this change, and will continue working through the other files 👍